### ISSUE TITLE
libavtools/avconv cannot be found on FreeBSD

### ISSUE THREAD

**Issue by niklaas** [Sonerezh/sonerezh#185]
I have `avconv` installed

```
$ command -v avconv                
/usr/local/bin/avconv
```

but `sonerezh` says that it is not available.

I guess the problem is that the check for libavtools/avconv is not POSIX compliant because it uses `which` and not `command -v`:

```
$ find . -type f -print0 | xargs -0 grep -n libavtools
./app/Controller/InstallersController.php:44:                   $libavtools = shell_exec("where avconv") || shell_exec("where ffmpeg");//WIN
./app/Controller/InstallersController.php:46:                   $libavtools = shell_exec("which avconv") || shell_exec("which ffmpeg");//NO WIN
./app/Controller/InstallersController.php:49:        if ($libavtools) {
./app/Controller/InstallersController.php:50:            $requirements['libavtools'] = array('label' => 'success', 'message' => __('libav-tools (avconv) is installed!'));
./app/Controller/InstallersController.php:52:            $requirements['libavtools'] = array('label' => 'warning', 'message' => __('libav-tools (avconv) is missing. Sonerezh will not be able to convert your tracks.'));
./app/Controller/InstallersController.php:202:            $this->request->data['Setting']['enable_auto_conv'] = $libavtools;
```

See also: http://stackoverflow.com/a/677212

**Comment by niklaas** [comment]
Hm. I did the following but could not solve the issue.

```
diff --git a/app/Controller/InstallersController.php b/app/Controller/InstallersController.php
old mode 100644
new mode 100755
index fcc8184..a546e6c
--- a/app/Controller/InstallersController.php
+++ b/app/Controller/InstallersController.php
@@ -43,7 +43,7 @@ class InstallersController extends AppController {
 		if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN') {
 			$libavtools = shell_exec("where avconv") || shell_exec("where ffmpeg");//WIN
 		} else {
-			$libavtools = shell_exec("which avconv") || shell_exec("which ffmpeg");//NO WIN
+			$libavtools = shell_exec("command -v avconv") || shell_exec("command -v ffmpeg");//NO WIN
 		}
 
         if ($libavtools) {
@@ -310,4 +310,4 @@ class InstallersController extends AppController {
         return $hash;
     }
 
-}
\ No newline at end of file
+}

```

**Comment by niklaas** [comment]
That's not it (although `command -v` remains more elegant than `which`). The problem is rather that `$PATH` of the `www` user on FreeBSD does not include `/usr/local/bin` where `avconv`lies. Thus, `shell_exec()` returns nothing. I got this idea from http://php.net/manual/en/function.shell-exec.php#78506.

I'm trying to solve this by globally including `/usr/local/bin/ in `www`'s `$PATH`. I'll keep you updated.

**Comment by niklaas** [comment]
Ok, so I've got some progress on this. $PATH is not set with `nginx` and `php-fpm` the way it needs to be for `sonerezh` to work. I refer to the submission I just made to one of FreeBSD's mailinglists: https://docs.freebsd.org/cgi/getmsg.cgi?fetch=513795+0+current/freebsd-questions. As explained there, I had to set the environment path I need in one of the pool configurations in `php-fpm.conf`.

Nonetheless, `sonerezh` doesn't find libavtools/avconv. I haven't further inspected that but

```
<?php                                                                                                                                            

$cmd = 'which avconv'; 
echo '<pre>' . shell_exec($cmd) . '</pre>';

?>
```

gives me `/usr/local/bin/avconv` now. So it is there at least.

**Comment by lGuillaume124** [comment]
Hi,
Very interesting :) I have never used FreeBSD so your suggestions are welcome!

**Comment by niklaas** [comment]
So, there is an update on this. In my previous post I wrote that it's still not working but that was an false alarm. I realised that changes that I made to `SettingsController.php` caused the errors. That said, what I wrote on the mailinglist fixes this issue.

I have no idea how to hardcode that solution into `sonerezh`'s code but a reference this issue and/or the entry of the mailinglist in `sonerezh`'s documentation should help future FreeBSDlers to run into this issue.

Note: This only applies to version 1.0.0. And I am not able to import music (nothing happens on a database update, no music is found). I get internal server errors with any version higher. So, what I will do now is try to get the most recent version running on FreeBSD. You'll get another bug report soon. ;-)
